src/lang/builtins_json.ml

```diff
@@ -21,62 +21,73 @@
  *****************************************************************************)
 
 let log = Log.make ["lang"; "json"]
-
-(* Utils.escape implements JSON's escaping RFC. *)
 let print_s s = Utils.escape_string (fun x -> Utils.escape_utf8 x) s
+let to_json_ref = ref (fun ~compact:_ _ -> assert false)
+
+module JSON = Lang.MkAbstract (struct
+  type content = (string, Lang.value) Hashtbl.t
+
+  let name = "json"
+  let descr _ = "json"
+
+  let to_json ~compact v =
+    if Hashtbl.length v > 0 then
+      !to_json_ref ~compact
+        (Lang.list
+           (Hashtbl.fold (fun k v l -> Lang.tuple [Lang.string k; v] :: l) v []))
+    else "{}"
+
+  let compare = Stdlib.compare
+end)
 
 let rec to_json_compact v =
   match v.Lang.value with
-    (* JSON specs do not allow a trailing . *)
-    | Lang.(Ground (Ground.Float n)) ->
-        let s = string_of_float n in
-        let s = Printf.sprintf "%s" s in
-        if s.[String.length s - 1] = '.' then Printf.sprintf "%s0" s else s
-    | Lang.(Ground (Ground.String s)) -> print_s s
-    | Lang.Ground g -> Lang_values.Ground.to_string g
+    | Lang.Ground g -> Lang_values.Ground.to_json ~compact:true g
     | Lang.List l -> (
-        try
-          (* Convert (string*'a) list to object *)
-          let l =
-            List.map
-              (fun x ->
-                let x, y = Lang.to_product x in
-                Printf.sprintf "%s:%s" (to_json_compact x) (to_json_compact y))
-              l
-          in
-          Printf.sprintf "{%s}" (String.concat "," l)
-        with _ ->
-          Printf.sprintf "[%s]" (String.concat "," (List.map to_json_compact l))
-        )
+        let v = String.concat "," (List.map to_json_compact l) in
+        match l with
+          | {
+              Lang.value =
+                Lang.Tuple
+                  [{ Lang.value = Lang.Ground (Lang.Ground.String _) }; _];
+            }
+            :: _ ->
+              Printf.sprintf "{%s}" v
+          | _ -> Printf.sprintf "[%s]" v)
```

Comment:
This should be removed now, it is unpredictable (empty `[string, 'a]` list will convert to `[]`) and is now superseded.